docs/enhancements/dual-stack-vips.md

```diff
@@ -0,0 +1,114 @@
+
+#### Phase 1
+
+`cluster-create-params`, `v2-cluster-update-params` and `cluster` will get new fields
+
+* `api_vips` of type `array`
+* `ingress_vips` of type `array`
+
+Respective AgentClusterInstall CR will also get those new fields.
+
+The database schema will be updated by creating new tables `api_vips` and `ingress_vips` with the following columns
+
+* `cluster_id`
+* `vip`
+
+#### Phase 2
+
+A set of functions for cluster creation will provide a compatiblity between already existing fields (`api_vip` and `ingress_vip`) and new ones (`api_vips` and `ingress_vips`). The behaviour will be as follows
+
+* If only `api_vip` is provided, the `api_vips` will be automatically populated by a simple conversion from `string` to `[]string`. This guarantees no change in behaviour for users not aware of introduction of the new field.
+
+* If both `api_vip` and `api_vips` are provided, the value of `api_vip[0]` must match the value of `api_vip`.
+
+* If only `api_vips` are provided, the error will be raised. This is the case when plural field is specified and the singular is not. The rationale is that in an update it's impossible to tell the difference between an old client clearing the singular field via patch and a new client setting the plural field. The update semantics must not differ from create, therefore behaviour as described.
+
+For cluster update the following logic will apply
+
+* If `api_vip` is cleared and `api_vips` are not provided, we will clear both fields. This is a scenario when an old client (not aware of plural field) is used.
+
+* If `api_vips` is cleared and `api_vip` is not provided, the change will not apply. This is a scenario when an old client is used and it can't send fields it doesn't know about.
+
+* If `api_vip` is changed (but not cleared), `api_vips` will be forcefully set to the value of `api_vip` as a one-element list. This is a scenario when an old client is used to change the field it knows about.
+
+All the points described above will behave in an equivalent way for `ingress_vip` and `ingress_vips`.
+
+#### Phase 3
+
+After providing a required period of backwards-compatibility the `api_vip` and `ingress_vip` fields will be deprecated and removed. The value in the database will be nulled. Removal of the column in the database will follow.
+
+### Risks and Mitigations
+
+The similar manouever of introducing `[]string` in place of `string` has already happened in the past when implementing the initial dual-stack support. An example of the field that got transformed is `machine_network_cidr` that became `machine_networks`.
+
+There is already an experience and established path, therefore the risk is considered minimal.
+
+Given that the core of the feature is implemented in the baremetal networking component of the OCP, the implementation in Assisted Installer is only responsible for providing the correct syntax to the install-config.
+
+### UI Impact
```

Comment:
@nirfarkas @jkilzi: please review and comment before any real-time kick-off meeting happens

Reply:
@nirfarkas @jkilzi this is a ping, I'm planning to proceed with the agreed implementation and I will be taking no comments as an agreement

Reply:
Questions:
1. How many entries per field (api_vips/ingress_vips) are users expected to enter?
2. Is it valid to pass IPv4 addresses intertwined with IPv6?
3. Is the order of the entries in the lists important?
4. WLG. Can users specify only IPv4 addresses in one list but mixed IP versions in the other?
5. Is there any relationship between the api_vips and ingress_vips, meaning, are the addresses required to be part of the same subnet when entered manually (no DHCP)?
6. Should we expect any special case for when VIP DHCP allocation is enabled/disabled?

Reply:
Answered 1-6 in goals and non-goals sections. Let me know if any of questions still needs some further clarification

Reply:
We still do need the following clarifications:
1. Is this enhancement document applicable if users opt to not use dual-stack?
1. Do users still need to enter multiple IPv4 addresses if they don't use dual-stack?
1. Goal: "Allow configuring only IPv4 VIPs for dual-stack clusters.". Isn't this something we already allow?
1. What kind of constrains should be applied to these fields?
1. Could you please provide some examples of valid or invalid user input?

Reply:
> How many entries per field (api_vips/ingress_vips) are users expected to enter?

This question is not addressed in the Goals section. Right now my understanding is that the UI should allow users to enter one IPv4 and one IPv6 per VIP when they choose to use dual-stack, is my assumption correct?

Reply:
Questions 1-2: This enhancement is only about dual-stack users. For people doing single-stack nothing changes. Hence, we don't mention them because we focus here only on people who do dual-stack.

Question 3: Yes you are right, we allow it now. It's in the goals explicitly so that it's clear that we will still allow it

Question 4: Explained in the doc. It's only about the belonging to the machine networks and about the order

>This question is not addressed in the Goals section. Right now my understanding is that the UI should allow users to enter one IPv4 and one IPv6 per VIP when they choose to use dual-stack, is my assumption correct?

It is addressed in the doc. We explain that enabling more than 1 VIP per IP stack is in the realm of MetalLB hence is not the goal of this enhancement here. TLDR is that you can provide 1 IPv4 and 1 IPv6, not more.